### PR TITLE
Thread-safe access to graph cache

### PR DESCRIPTION
**Pull request by Yadunund**
Step 1 in addressing #249.

This PR updates the `rmw_context_impl_s` class to accurately manage the lifetime of its members while ensure thread-safe data access. For starters, it ensures that the Graph Cache updates and lookups are thread safe.

In subsequent PRs, I will update the `rmw_context_impl_s` class to store `rmw_publisher_data_t`, etc with member functions to manage/access their functionalities.

### CODE REVIEW

File: rmw_zenoh_cpp/src/rmw_init.cpp
Comment (clalancette) [inline rmw_zenoh_cpp/src/rmw_init.cpp]: In the new structure, it seems like the `rmw_context_impl_s` structure is the "owner" of the `graph_guard_condition`.  If I'm correct about that, then my suggestion here is to move the "construction" of the graph guard_condition (this code, and the next dozen lines or so) into the constructor of `rmw_context_impl_s`.  I think it makes more sense there.  What do you think?

File: rmw_zenoh_cpp/src/rmw_zenoh.cpp
Comment (clalancette) [inline rmw_zenoh_cpp/src/rmw_zenoh.cpp:172]: I don't have a good idea of whether or not this check should be in place.  What was your reasoning for removing it?

File: rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp
Comment (clalancette) [inline rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp:334]: This lock isn't doing too much.  The problem is that as soon as the call to `rmw_context_impl_s::session()` returns, the lock is dropped.  At that point, the caller has a loan to the session, but if some other thread changes the session, the caller now has a dangling loan.  We'll need to do something more clever with sessions here in order to make it thread safe.

This same comment applies to the `shm_manager()` and `graph_guard_condition()` methods below (it doesn't apply to `enclave()` above because that is returning a copy).

**Review by clalancette — COMMENTED** [review]
I've left a few things to think about.

File: rmw_zenoh_cpp/src/rmw_init.cpp
Comment (Yadunund) [reply rmw_zenoh_cpp/src/rmw_init.cpp]: Yeah I wanted to do that  but was unsure how to handle the case when we need to return `RMW_RET_BAD_ALLOC` when memory allocation fails. Should the constructor throw an error? In #259 I was able to rely on the `make()` pattern but I don't see a way to do that here unless we switch to a singleton of `rmw_context_impl_s`.

File: rmw_zenoh_cpp/src/rmw_zenoh.cpp
Comment (Yadunund) [reply rmw_zenoh_cpp/src/rmw_zenoh.cpp:172]: Ah I now store the `enclave` as an `std::string` within `rmw_context_impl_s` so we don't need to check for nullptr. Maybe I can update to check if the string is empty?

File: rmw_zenoh_cpp/src/rmw_init.cpp
Comment (clalancette) [reply rmw_zenoh_cpp/src/rmw_init.cpp]: I think the constructor can throw; we are careful to use `RMW_TRY_PLACEMENT_NEW` when constructing, which captures the exception and turns it into a `return` through the C API.

File: rmw_zenoh_cpp/src/rmw_zenoh.cpp
Comment (clalancette) [reply rmw_zenoh_cpp/src/rmw_zenoh.cpp:172]: Oh, I see.  Right, that makes sense.

I kind of feel like an empty enclave is allowed (when not using security).  In that case, I think what you have here is fine.

File: rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp
Comment (Yadunund) [reply rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp:334]: Yes I'm aware of this issue and this is a temporary workaround (I left a similar comment in the function that returns the `std::optional<shm_manager>`).

My goal in subsequent PRs is to add functions like `rmw_context_impl_s::add_publisher()`, `rmw_context_impl_s::publish()` etc which will create/handle the publisher/subscribption/client/service instance inside `rmw_context_impl_s` directly without returning anything. This way we will lock the `session` -> create the instance -> add it to an internal map like in #259.

File: rmw_zenoh_cpp/src/rmw_init.cpp
Comment (Yadunund) [reply rmw_zenoh_cpp/src/rmw_init.cpp]: Ah that's pretty neat! Will update accordingly!

File: rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp
Comment (Yadunund) [reply rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp:334]: ie this comment https://github.com/ros2/rmw_zenoh/blob/a9b34e2ede699450cc5dd968e87f172827cdbb28/rmw_zenoh_cpp/src/detail/rmw_context_impl_s.hpp#L54-L58

File: rmw_zenoh_cpp/src/rmw_init.cpp
Comment (Yadunund) [reply rmw_zenoh_cpp/src/rmw_init.cpp]: After learning that `RMW_TRY_PLACEMENT_NEW` elegantly handles exceptions, I've updated the constructor here to also instantiate the Zenoh router and shm manager. https://github.com/ros2/rmw_zenoh/pull/258/commits/a6fb52136f499dd411de284a912ef0cd99b2e85f

File: rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp
Comment (clalancette) [inline rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp]: Isn't this equivalent to `this`?

File: rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp
Comment (clalancette) [inline rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp]: I think the method name here could be a little more descriptive.  `subscribe_to_graph_cache`, maybe?

File: rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp
Comment (clalancette) [inline rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp:107]: Do we need to take the `mutex_`, for guarding access to `is_initialized_` and `graph_cache_`?

File: rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp
Comment (clalancette) [inline rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp]: We might consider getting rid of this check; if we've gotten this far, then the `allocator_` had to have been valid, right?  Maybe we should move this check into the constructor instead?

File: rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp
Comment (clalancette) [inline rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp]: I know you are just porting this over, but I'm just realizing that one second is a long time to wait here.  I would much rather see us check more often, like every 100 milliseconds or so.  I don't think we should fix it in this PR, but I'm going to suggest opening an issue (or adding a comment to an existing issue, if one is appropriate) to reconsider this.

**Comment by MichaelOrlov** [comment]
@Yadunund Friendly ping to follow up on this issue.

File: rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp
Comment (ahcorde) [inline rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp:57]: #include `<mutex>`

**Comment by MichaelOrlov** [comment]
Discussion from maintenance triage: Decided to assign this issue to @Yadunund

**Comment by clalancette** [comment]
> `rmw_context_impl_s::publish()` etc

I think this is what I'm struggling with most in this change.  It just doesn't seem right to me to be having one giant class that encapsulates all functionality of the RMW.  It does fix the locking problem, but it doesn't seem very elegant.

If we ignore the threading problem for the moment, in my ideal world we'd have a `ContextImpl` class at the top-level.  During `rmw_create_node`, we'd call `ContextImpl::create_node()`, which would return a `Node` object.  During `rmw_create_publisher`, we'd call `Node::create_publisher()`, which would return a `Publisher` object.  During `rmw_publish`, we'd call `Publisher::publish()`.  (There would be similar classes for `Subscription`, `ServiceClient`, and `ServiceServer`).  And the `GraphCache` would be embedded inside of the `ContextImpl` (since that is indeed a session-wide entity).  In that class hierarchy, the functionality for each of these entities is encapsulated into its own class, and further they don't have to share one giant lock.  This is, incidentally, how zenoh-cpp works (though there is no `Node` entity there, so you ask the `Session` to create a `Publisher` object).

Now we have to think about the locking.  If we just did the above, it wouldn't be much different from what we have today, and it wouldn't fix our locking problems.  So what we need is for each entity to be able to query its "parent" on whether it is still alive.  For instance, in `rmw_publish`, we'd have to ask the `Node` whether the `Publisher` is still alive.  But before that, we have to actually ask the `ContextImpl` whether the `Node` is still alive.  (There's an argument to be made here that we have a similar problem with `ContextImpl`, but I'm not sure we can solve that with the current RMW API).

Thus, the `Node` class would keep a list of active `Publisher`, `Subscription`, `ServiceServer`, and `ServiceClient` objects.  The `ContextImpl` class would keep a list of active `Node` objects.  And the opaque data (`rmw_publisher->data`, etc) that we return from `rmw_create_publisher` (and siblings) would be a pointer to the `ContextImpl`, some kind of identifier for the `Node`, and some kind of identifier for the `Publisher`.  Then we'd have enough information to walk the entire hierarchy and determine whether the thing we want is still alive.  During an `rmw_destroy_publisher`, we'd remove it from the `Node` list, and thus further calls wouldn't work.  `rmw_destroy_subscription` is a bit harder, because a callback may race here; we'd have to make the callback ask the node whether the subscription is still alive at this point.

I admit I haven't totally thought through all of the implications here.  But I think this proposal creates a class hierarchy, but also allows us to add in additional checking/locking.  What do you think of the general idea?

**Comment by Yadunund** [comment]
> > `rmw_context_impl_s::publish()` etc
> 
> I think this is what I'm struggling with most in this change. It just doesn't seem right to me to be having one giant class that encapsulates all functionality of the RMW. It does fix the locking problem, but it doesn't seem very elegant.
> 
> If we ignore the threading problem for the moment, in my ideal world we'd have a `ContextImpl` class at the top-level. During `rmw_create_node`, we'd call `ContextImpl::create_node()`, which would return a `Node` object. During `rmw_create_publisher`, we'd call `Node::create_publisher()`, which would return a `Publisher` object. During `rmw_publish`, we'd call `Publisher::publish()`. (There would be similar classes for `Subscription`, `ServiceClient`, and `ServiceServer`). And the `GraphCache` would be embedded inside of the `ContextImpl` (since that is indeed a session-wide entity). In that class hierarchy, the functionality for each of these entities is encapsulated into its own class, and further they don't have to share one giant lock. This is, incidentally, how zenoh-cpp works (though there is no `Node` entity there, so you ask the `Session` to create a `Publisher` object).
> 
> Now we have to think about the locking. If we just did the above, it wouldn't be much different from what we have today, and it wouldn't fix our locking problems. So what we need is for each entity to be able to query its "parent" on whether it is still alive. For instance, in `rmw_publish`, we'd have to ask the `Node` whether the `Publisher` is still alive. But before that, we have to actually ask the `ContextImpl` whether the `Node` is still alive. (There's an argument to be made here that we have a similar problem with `ContextImpl`, but I'm not sure we can solve that with the current RMW API).
> 
> Thus, the `Node` class would keep a list of active `Publisher`, `Subscription`, `ServiceServer`, and `ServiceClient` objects. The `ContextImpl` class would keep a list of active `Node` objects. And the opaque data (`rmw_publisher->data`, etc) that we return from `rmw_create_publisher` (and siblings) would be a pointer to the `ContextImpl`, some kind of identifier for the `Node`, and some kind of identifier for the `Publisher`. Then we'd have enough information to walk the entire hierarchy and determine whether the thing we want is still alive. During an `rmw_destroy_publisher`, we'd remove it from the `Node` list, and thus further calls wouldn't work. `rmw_destroy_subscription` is a bit harder, because a callback may race here; we'd have to make the callback ask the node whether the subscription is still alive at this point.
> 
> I admit I haven't totally thought through all of the implications here. But I think this proposal creates a class hierarchy, but also allows us to add in additional checking/locking. What do you think of the general idea?

Thanks for the additional thoughts here. I agree such a hierarchy would be ideal. What i'm not clear about is whether `Node` will have an API to return a `Publisher::SharedPtr`? My only concern here is that we need to ensure that the only thing that `rmw_destroy_publisher` does is remove the `Publisher::SharedPtr` from the container in which it is stored within `Node`. Then the destructor of `Publisher` takes care of all the cleaning up. So even if `rmw_publish` obtained this `Publisher::SharedPtr` from `Node`, and is thus keeping it alive while `rmw_destroy_publisher` is invoked in a separate thread, the `Publisher's` destructor will only be invoked after `rmw_publish` returns. This would also be the case if `Node` is destroyed  when `rmw_publish` is being invoked.

In any case it seems like this is a bridge we will need to cross only after we merge this PR; perhaps when addressing #259? So should we move the discussion there since the changes here simple make `rmw_context_impl_s` a concrete class with private data members?

**Comment by Yadunund** [comment]
In https://github.com/ros2/rmw_zenoh/pull/258/commits/7cc52bdb76ea9f4f8fb9b290483d1073b78a88d9, I updated the `rmw_context_impl_s` class to return a `shared_ptr<GraphCache>`. The `graph_sub_data_handler()` cb should still be thread-safe with the new implementation as the callback and other methods that query the `shared_ptr<GraphCache>` should lock the same mutex within `GraphCache`. 
Hoping this addresses the concern of `rmw_context_impl_s` class embedding everything within it. I will update #259 in a similar way such that `rmw_context_impl_s` returns a `shared_ptr<NodeData>`.

File: rmw_zenoh_cpp/src/detail/rmw_context_impl_s.hpp
Comment (ahcorde) [inline rmw_zenoh_cpp/src/detail/rmw_context_impl_s.hpp:44]: include `<cstddef>`

**Review by ahcorde — CHANGES_REQUESTED** [review]

File: rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp
Comment (Yadunund) [reply rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp:57]: The header is included in `rmw_context_impl_s.hpp`

File: rmw_zenoh_cpp/src/detail/rmw_context_impl_s.hpp
Comment (Yadunund) [reply rmw_zenoh_cpp/src/detail/rmw_context_impl_s.hpp:44]: Since we're compiling with C++, `std::size_t` is part of the standard library and should not need any additional includes. Please correct me if i'm mistaken.

File: rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp
Comment (clalancette) [reply rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp:57]: While true, I am always of the opinion that you should include-what-you-use.  So I'd add it in here as well.

File: rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp
Comment (clalancette) [inline rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp]:
```suggestion
#include <mutex>
#include <thread>
#include <utility>
```

File: rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp
Comment (Yadunund) [reply rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp]: https://github.com/ros2/rmw_zenoh/issues/277

File: rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp
Comment (Yadunund) [reply rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp]: When I move `allocator_->deallocate(graph_guard_condition_->data, allocator_->state);` outside of this check, i get a segfault when running some of the rcl tests. Here's the command to run the test directly (you will need to update the path to your rolling workspace)

```
/usr/bin/python3 "-u" "home/yadunund/ros2_rolling/install/ament_cmake_test/share/ament_cmake_test/cmake/run_test.py" "/home/yadunund/ros2_rolling/build/rcl/test_results/rcl/test_info_by_topic__rmw_zenoh_cpp.gtest.xml" "--package-name" "rcl" "--output-file" "/home/yadunund/ros2_rolling/build/rcl/ament_cmake_gtest/test_info_by_topic__rmw_zenoh_cpp.txt" "--env" "RMW_IMPLEMENTATION=rmw_zenoh_cpp" "--command" "/home/yadunund/ros2_rolling/build/rcl/test/test_info_by_topic" "--gtest_output=xml:/home/yadunund/ros2_rolling/build/rcl/test_results/rcl/test_info_by_topic__rmw_zenoh_cpp.gtest.xml"

```

Does this imply the allocator is being destroyed sooner than expected somewhere in the rcl layer?

File: rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp
Comment (Yadunund) [reply rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp:107]: Done https://github.com/ros2/rmw_zenoh/pull/258/commits/6ad087c383d5b164869b7a5a3f1f52247722e5a4

I didn't include it initially as I thought we might run into a deadlock between this thread that calls `zc_liveliness_declare_subscriber` and the Zenoh thread which might subscribe to a liveliness token and then trigger `rmw_context_impl_s::graph_sub_data_handler()` callback which also locks the same mutex.

File: rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp
Comment (Yadunund) [reply rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp]: Updated to `subscribe_to_ros_graph` https://github.com/ros2/rmw_zenoh/pull/258/commits/6ad087c383d5b164869b7a5a3f1f52247722e5a4

File: rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp
Comment (Yadunund) [reply rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp]: It totally is. I forgot to switch it back to `this` after playing around with `shared/weak_from_this`
https://github.com/ros2/rmw_zenoh/pull/258/commits/6ad087c383d5b164869b7a5a3f1f52247722e5a4

File: rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp
Comment (Yadunund) [reply rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp:57]: Done https://github.com/ros2/rmw_zenoh/pull/258/commits/6ad087c383d5b164869b7a5a3f1f52247722e5a4

File: rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp
Comment (clalancette) [reply rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp]: > Does this imply the allocator is being destroyed sooner than expected somewhere in the rcl layer?

Yeah, it has to be.

That said, looking at this more closely, this has the problem that I had elsewhere, in that "another entity" allocated these, and then the "Data" class is deallocating them.  While that can work if you are careful about transferring ownership, I think it is just simpler in general to have the entity that allocated them deallocate them (no transfer of ownership).  With that in mind, I'll suggest either moving the construction of the graph_guard_condition into the Data constructor, *or* moving the destruction here into the rmw_context_impl_s destructor.

Comment (Yadunund) [reply rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp]: I ended up not using the `allocator` when instantiating the `graph_guard_condition` https://github.com/ros2/rmw_zenoh/pull/258/commits/5be2532930a7ccefcf761811a601a3c3297a065f.

File: rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp
Comment (ahcorde) [inline rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp]:
```suggestion
    throw std::runtime_error("Error setting up zenoh session.");
```

File: rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp
Comment (ahcorde) [inline rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp]:
```suggestion
        "rmw_zenoh_cpp", "[rmw_context_impl_s] z_call received an invalid reply.\n");
```

File: rmw_zenoh_cpp/src/detail/rmw_context_impl_s.hpp
Comment (ahcorde) [inline rmw_zenoh_cpp/src/detail/rmw_context_impl_s.hpp:65]: include `<cstddef>`

File: rmw_zenoh_cpp/src/detail/rmw_context_impl_s.hpp
Comment (ahcorde) [inline rmw_zenoh_cpp/src/detail/rmw_context_impl_s.hpp]:
```suggestion
    // Shutdown flag.
```

File: rmw_zenoh_cpp/src/detail/rmw_context_impl_s.hpp
Comment (ahcorde) [inline rmw_zenoh_cpp/src/detail/rmw_context_impl_s.hpp]:
```suggestion
    // Guard condition that should be triggered when the graph changes.
```

**Review by ahcorde — APPROVED** [review]
Some minor fixes, otherwise LGTM

**Review by clalancette — APPROVED** [review]
This needs a rebase to fix conflicts, and we should fix @ahcorde 's comments, but this otherwise looks good to me.

**Pull request merged**
